.. contents::
   :depth: 3.0
..

Going forward
=============

Look into
---------

-  `Titus Brown and Colleague's Next-Gen Sequence Analysis
   Workshops <http://ged.msu.edu/angus/>`__, most recent is [Next-Gen
   Sequence Analysis Workshop
   (2014)(http://angus.readthedocs.org/en/2014/) Particularly pertinent
   are the sections `Istvan Albert's 2012 ChIP-Seq
   lecture <http://ged.msu.edu/angus/tutorials-2012/day5.html>`__, `Day
   7: ChIP-seq: Peak Predictions and Cis-regulatory Element
   Annotations <http://ged.msu.edu/angus/tutorials-2011/day7.html>`__,
   `Using MEME to identify TF binding motif from ChIP-seq
   data <http://ged.msu.edu/angus/tutorials/chipseq-motif-finding.html>`__
   and `here <http://ged.msu.edu/angus/tutorials-2012/day5.html>`__. For
   those ready for more see Titus's coverage, with many links, of his
   advanced week of classes offered first in 2015
   `here <http://ivory.idyll.org/blog/2015-small-batch.html>`__.

-  `ChIP- and DNase-seq data analysis workshop
   2014 <http://web.csc.fi/english/csc/courses/archive/chipseq2014>`__

-  `Cistrome <http://cistrome.org/Cistrome/Cistrome_Project.html>`__ - a
   web-based ChIP-seq analysis portal based on Galaxy open source
   framework run by the labs of Myles Brown and Xiaole Shirley Liu. It
   provides a complete workflow for ChIP-seq and downstram analysis
   without the need for local installation and configuration.

-  `A History of Bioinformatics (in the Year 2039), a presentation by
   Titus Brown encouraging good data practices in his unique
   way <https://www.youtube.com/watch?v=uwsjwMO-TEA>`__

File Formats and Conversion help
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

-  `UCSC Genome Bioinformatics- Data File Formats FAQ
   section <https://genome.ucsc.edu/FAQ/FAQformat.html>`__

-  `Data File format on Abecasis Group
   Wiki <http://genome.sph.umich.edu/wiki/Introduction_to_new_students#Data_File_Formats>`__,
   for example `SAM <http://genome.sph.umich.edu/wiki/SAM>`__

-  `Ensembl's documentation on file
   formats <http://useast.ensembl.org/info/website/upload/index.html#formats>`__

-  `Bam2Wig
   example <http://biostar.stackexchange.com/questions/2704/how-can-i-convert-bam-sam-to-wiggle>`__

-  `File Formats Listing associated with the Broad Institute's
   Integrated Genomics
   Viewer <https://www.broadinstitute.org/software/igv/FileFormats>`__

ChIP-seq data analysis
~~~~~~~~~~~~~~~~~~~~~~

-  `README for MACS (2.1.0) <https://github.com/taoliu/MACS/>`__

-  `Feng et al. 2012. Identifying ChIP-seq enrichment using MACS.Nat
   Protoc. 2012 Sep; 7(9):
   10.1038/nprot.2012.101. <http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3868217/>`__

-  `Using MACS to Identify Peaks from ChIP-Seq Data. Curr Protoc
   Bioinformatics. 2011 Jun; CHAPTER:
   Unit2.14. <http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3120977/>`__

-  `ChIP-Seq: technical considerations for obtaining high-quality data.
   Kidder BL, Hu G, Zhao K. Nat Immunol. 2011 Sep 20;12(10):918-22. doi:
   10.1038/ni.2117. PMID:
   21934668 <http://www.ncbi.nlm.nih.gov/pubmed/21934668>`__

-  `Titus Brown and Colleague's Next-Gen Sequence Analysis
   Workshops <http://ged.msu.edu/angus/>`__, most recent is [Next-Gen
   Sequence Analysis Workshop
   (2014)(http://angus.readthedocs.org/en/2014/) Particularly pertinent
   are the sections `Istvan Albert's 2012 ChIP-Seq
   lecture <http://ged.msu.edu/angus/tutorials-2012/day5.html>`__, `Day
   7: ChIP-seq: Peak Predictions and Cis-regulatory Element
   Annotations <http://ged.msu.edu/angus/tutorials-2011/day7.html>`__,
   `Using MEME to identify TF binding motif from ChIP-seq
   data <http://ged.msu.edu/angus/tutorials/chipseq-motif-finding.html>`__
   and `here <http://ged.msu.edu/angus/tutorials-2012/day5.html>`__.

-  `ChIP-seq analysis using GALAXY: Part 1 -
   Introduction <https://www.youtube.com/watch?v=2ehtAVqwmNU>`__ -uses
   MACS, CEAS, and WEEDER. The latter being an alternative to MEME for
   motif discovery.

-  `The ChIP-seq pipeline “Cistrome” integrated into
   UGENE <https://ugene.unipro.ru/wiki/display/WDD/ChIP-seq+Analysis+with+Cistrome+Tools>`__
   uses MACS and CEAS and additional characterization of binding sites.
   You can read about it and the UGENE NGS platform
   `here <https://peerj.com/articles/644/>`__. The UGENE pipeline is
   originally based on the General ChIP-seq pipeline from the public
   `Cistrome <http://cistrome.org/Cistrome/Cistrome_Project.html>`__
   installation on the Galaxy workflow platform

-  `ChIP- and DNase-seq data analysis workshop
   2014 <http://web.csc.fi/english/csc/courses/archive/chipseq2014>`__

-  `Cis-regulatory Element Annotation
   System <http://liulab.dfci.harvard.edu/CEAS/>`__ by Hyunjin Shin and
   Tao Liu from Xiaole Shirley Liu's Lab

-  `ab initio motif finder
   MEME <http://www.ncbi.nlm.nih.gov/pubmed/16845028>`__ and the related
   `MEME suite <http://www.ncbi.nlm.nih.gov/pubmed/25953851>`__

-  `MEME-LaB wraps the popular ab initio motif finder in a web
   tool <http://www.ncbi.nlm.nih.gov/pubmed/23681125>`__

-  `Motif enrichment tool. Blatti C, Sinha S. Nucleic Acids Res. 2014
   Jul;42(Web Server issue):W20-5. doi: 10.1093/nar/gku456. Epub 2014
   May 23. PMID:
   24860165 <http://www.ncbi.nlm.nih.gov/pubmed/24860165>`__

-  `Motif-based analysis of large nucleotide data sets using
   MEME-ChIP <http://www.ncbi.nlm.nih.gov/pubmed/24853928>`__

ChIP-exo methods
~~~~~~~~~~~~~~~~

-  `Rhee HS and Pugh BF (2011) Comprehensive Genome-wide Protein-DNA
   Interactions Detected at Single-Nucleotide Resolution. Cell
   147(6):1408-19 <http://www.ncbi.nlm.nih.gov/pubmed/22153082>`__

-  `Several groups report good results and Illumina and other companies
   have kits and
   protocols <http://seqanswers.com/forums/showthread.php?t=38980>`__

-  `Development of an Illumina-based ChIP-exonuclease method provides
   insight into FoxA1-DNA binding
   properties <http://www.ncbi.nlm.nih.gov/pubmed/24373287>`__

-  `Webinar: “Precision Mapping”- High Resolution Mapping using
   ChIP-exo" <http://epigenie.com/webinar-precision-mapping-high-resolution-mapping-using-chip-exo/>`__

-  `No formaldehyde, no problem for new genome-wide ChIP
   method <https://www.fredhutch.org/en/news/spotlight/imports/no-formaldehyde--no-problem-for-new-genome-wide-chip-method.html>`__

    ORGANIC detected more Reb1 binding sites than either X-ChIP-chip or
    ChIP-exo.

Questions
~~~~~~~~~

-  Try Google, probably will lead you to one of my listed resources
   or...
-  `Biostars <https://www.biostars.org>`__
-  Stackoverflow for general scripting and computing
-  `SEQanswers <http://seqanswers.com/>`__ - a high throughput
   sequencing community

-  Try Twitter - for example
   `this <https://twitter.com/jaredtsimpson/status/535443211452702720>`__
