.. contents::
   :depth: 3.0
..

Sources
=======

The sources for the information used today came from those linked
throughout the content.

However, certain sources deserve special highlighting as they were
particularly useful in developing this presentation, contain a wealth of
related resources, or are especially pertinent at this stage.

-  `README for MACS (2.1.0) <https://github.com/taoliu/MACS/>`__

-  `Feng et al. 2012. Identifying ChIP-seq enrichment using MACS.Nat
   Protoc. 2012 Sep; 7(9):
   10.1038/nprot.2012.101. <http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3868217/>`__

-  `Using MACS to Identify Peaks from ChIP-Seq Data. Curr Protoc
   Bioinformatics. 2011 Jun; CHAPTER:
   Unit2.14. <http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3120977/>`__

-  

   -  `ChIP-Seq: technical considerations for obtaining high-quality
      data. Kidder BL, Hu G, Zhao K. Nat Immunol. 2011 Sep
      20;12(10):918-22. doi: 10.1038/ni.2117. PMID:
      21934668 <http://www.ncbi.nlm.nih.gov/pubmed/21934668>`__

-  `Titus Brown and Colleague's Next-Gen Sequence Analysis
   Workshops <http://ged.msu.edu/angus/>`__, most recent is [Next-Gen
   Sequence Analysis Workshop
   (2014)(http://angus.readthedocs.org/en/2014/) Particularly pertinent
   are:

   -  the sections `Istvan Albert's 2012 ChIP-Seq
      lecture <http://ged.msu.edu/angus/tutorials-2012/day5.html>`__
   -  `Day 7: ChIP-seq: Peak Predictions and Cis-regulatory Element
      Annotations <http://ged.msu.edu/angus/tutorials-2011/day7.html>`__\ \*
      `Using MEME to identify TF binding motif from ChIP-seq
      data <http://ged.msu.edu/angus/tutorials/chipseq-motif-finding.html>`__
      \* `here <http://ged.msu.edu/angus/tutorials-2012/day5.html>`__.
   -  For the mapping of the reads using bwa I relied heavily on
      `Mapping reads with bwa and bowtie
      tutorial <http://ged.msu.edu/angus/tutorials-2011/bwa_tutorial.html>`__
   -  `Software packages to
      install <http://ged.msu.edu/angus/2013-04-assembly-workshop/installing-software.html>`__
   -  Another recent `guide to software
      installation <http://davis-assembly-masterclass-2013.readthedocs.org/en/latest/lex-notes.html#installing-software>`__

-  `ChIP- and DNase-seq data analysis workshop
   2014 <http://web.csc.fi/english/csc/courses/archive/chipseq2014>`__

-  `UCSC Genome Bioinformatics- Data File Formats FAQ
   section <https://genome.ucsc.edu/FAQ/FAQformat.html>`__

-  `Rhee HS and Pugh BF (2011) Comprehensive Genome-wide Protein-DNA
   Interactions Detected at Single-Nucleotide Resolution. Cell
   147(6):1408-19 <http://www.ncbi.nlm.nih.gov/pubmed/22153082>`__
